File: wun-android/src/main/kotlin/wun/foundation/StackRenderer.kt
// Compose renderer for `:wun/Stack`. Direction defaults to column;
// `:row` switches to Row. `:gap` and `:padding` are pixel ints.

package wun.foundation

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive
import wun.WunComponent
import wun.WunView

object StackRenderer {
    val render: WunComponent = renderer@ { props, children ->
        val direction = (props["direction"] as? JsonPrimitive)?.let {
            if (it.isString) it.content else null
        }
        val gap     = (props["gap"]?.jsonPrimitive?.intOrNull) ?: 0
        val padding = (props["padding"]?.jsonPrimitive?.intOrNull) ?: 0
        val mod = Modifier.padding(padding.dp)
        if (direction == "row") {
            Row(modifier = mod,
                horizontalArrangement = Arrangement.spacedBy(gap.dp),
                verticalAlignment = Alignment.CenterVertically) {
                children.forEach { kid -> WunView(kid) }
            }
        } else {
            Column(modifier = mod,
                   verticalArrangement = Arrangement.spacedBy(gap.dp)) {
                children.forEach { kid -> WunView(kid) }
            }
        }
    }
}

File: wun-ios/Sources/Wun/Foundation/StackRenderer.swift
// SwiftUI renderer for `:wun/Stack`. Direction defaults to column;
// `:row` switches to HStack. `:gap` and `:padding` are pixel ints.
//
// User code registers the same way:
//   registry.register("myapp/Card", MyAppCard.render)

import SwiftUI

public enum WunStack {
    public static let render: WunComponent = { props, children in
        let direction = props["direction"]?.stringValue ?? "column"
        let gap       = (props["gap"]?.intValue).map { CGFloat($0) }     ?? 0
        let pad       = (props["padding"]?.intValue).map { CGFloat($0) } ?? 0

        return AnyView(
            Group {
                if direction == "row" {
                    HStack(alignment: .center, spacing: gap) {
                        ForEach(Array(children.enumerated()), id: \.offset) { _, kid in
                            WunView(kid)
                        }
                    }
                } else {
                    VStack(alignment: .leading, spacing: gap) {
                        ForEach(Array(children.enumerated()), id: \.offset) { _, kid in
                            WunView(kid)
                        }
                    }
                }
            }
            .padding(pad)
        )
    }
}
